src/registry.c

```diff
@@ -0,0 +1,1196 @@
+        return false;
+    }
+
+    tmp = strdup(path);
+    if (!tmp)
+        goto err;
+
+    err = stat(path, &stat_buf);
+    if (err != 0)
+        goto err;
+    if (!S_ISDIR(stat_buf.st_mode))
+        goto err;
+
+    if (!check_eaccess(path, R_OK | X_OK))
+        goto err;
+
+    /* Pre-filter for the 99.9% case - if we can't assemble the default ruleset
+     * path, complain here instead of during parsing later. The niche cases
+     * where this is the wrong behaviour aren't worth worrying about.
+     */
+    if (!snprintf_safe(rules, sizeof(rules), "%s/rules/%s.xml",
+                       path, DEFAULT_XKB_RULES))
+        goto err;
+
+    darray_append(ctx->includes, tmp);
+
+    return true;
+
+err:
+    free(tmp);
+    return false;
+}
+
+XKB_EXPORT bool
+rxkb_context_include_path_append_default(struct rxkb_context *ctx)
+{
+    const char *home, *xdg, *root;
+    char *user_path;
+    int err;
+    bool ret = false;
+
+    if (ctx->context_state != CONTEXT_NEW) {
+        log_err(ctx, "include paths can only be appended to a new context\n");
+        return false;
+    }
+
+    home = secure_getenv("HOME");
+
+    xdg = secure_getenv("XDG_CONFIG_HOME");
+    if (xdg != NULL) {
+        err = asprintf(&user_path, "%s/xkb", xdg);
+        if (err >= 0) {
+            ret |= rxkb_context_include_path_append(ctx, user_path);
+            free(user_path);
+        }
+    } else if (home != NULL) {
+        /* XDG_CONFIG_HOME fallback is $HOME/.config/ */
+        err = asprintf(&user_path, "%s/.config/xkb", home);
+        if (err >= 0) {
+            ret |= rxkb_context_include_path_append(ctx, user_path);
+            free(user_path);
+        }
+    }
+
+    if (home != NULL) {
+        err = asprintf(&user_path, "%s/.xkb", home);
+        if (err >= 0) {
+            ret |= rxkb_context_include_path_append(ctx, user_path);
+            free(user_path);
+        }
+    }
+
+    root = secure_getenv("XKB_CONFIG_ROOT");
+    if (root != NULL)
+        ret |= rxkb_context_include_path_append(ctx, root);
+    else
+        ret |= rxkb_context_include_path_append(ctx, DFLT_XKB_CONFIG_ROOT);
+
+    return ret;
+}
+
+XKB_EXPORT bool
+rxkb_context_parse_default_ruleset(struct rxkb_context *ctx)
+{
+    return rxkb_context_parse(ctx, DEFAULT_XKB_RULES);
+}
+
+XKB_EXPORT bool
+rxkb_context_parse(struct rxkb_context *ctx, const char *ruleset)
+{
+    char **path;
+    bool success = false;
+
+    if (ctx->context_state != CONTEXT_NEW) {
+        log_err(ctx, "parse must only be called on a new context\n");
+        return false;
+    }
+
+    darray_foreach_reverse(path, ctx->includes) {
+        char rules[PATH_MAX];
+
+        if (snprintf_safe(rules, sizeof(rules), "%s/rules/%s.xml",
+                           *path, ruleset)) {
+            log_dbg(ctx, "Parsing %s\n", rules);
+            if (parse(ctx, rules, RXKB_POPULARITY_STANDARD))
+                success = true;
+        }
```

Comment:
Maybe log an error if the snprintf fails?

Reply:
Do we need to worry about a use-case where we exceed `PATH_MAX` for file length? I'm pretty sure all sorts of funny issues happen then. For the default case (system + `XDG_CONFIG_HOME`) we should never hit this and `rxkb_context_include_path_append()` has a check for this case so it won't let you append the path to begin with. There are some cases to get around it but IMO they're not worth worrying about.